MekHQ/src/mekhq/campaign/autoresolve/acar/handler/StandardUnitAttackHandler.java

```diff
@@ -0,0 +1,231 @@
+
+                if (newArmor * 2 <= targetUnit.getCurrentArmor()) {
+                    target.setHighStressEpisode();
+                    reporter.reportStressEpisode();
+                }
+
+                targetUnit.setCurrentArmor(newArmor);
+
+                if (target.isCrippled() && newArmor > 0) {
+                    reporter.reportStressEpisode();
+                    target.setHighStressEpisode();
+                    reporter.reportUnitCrippled();
+                }
+
+                if (newArmor == 0) {
+                    // Destroyed
+                    reporter.reportUnitDestroyed();
+                    target.setHighStressEpisode();
+                    countKill(attackingUnit, targetUnit);
+                } else {
+                    // Check for critical hits if armor is now less than half original
+                    if (newArmor * 2 < targetUnit.getArmor()) {
+                        reporter.reportCriticalCheck();
+                        var critRoll = Compute.rollD6(2);
+                        var criticalRollResult = critRoll.getIntValue();
+                        handleCrits(target, targetUnit, criticalRollResult, attackingUnit);
+                    }
+                }
+            }
+        }
+    }
+
+    private int calculateDamage(Formation attacker, StandardUnitAttack attack,
+                                SBFUnit attackingUnit, Formation target) {
+        int damage = 0;
+        if (attack.getManeuverResult().equals(StandardUnitAttack.ManeuverResult.SUCCESS)) {
+            damage += 1;
+        }
+        damage += attackingUnit.getCurrentDamage().getDamage(attack.getRange()).damage;
+
+        damage = Math.max(1, processDamageByEngagementControl(attacker, target, damage));
+        return damage;
+    }
+
+    private void handleCrits(Formation target, SBFUnit targetUnit, int criticalRollResult, SBFUnit attackingUnit) {
+        if (criticalRollResult <= 4) {
+            // No crit
+            reporter.reportNoCrit();
+        } else if (criticalRollResult <= 7) {
+            // targeting critical
+            targetUnit.addTargetingCrit();
+            reporter.reportTargetingCrit(targetUnit);
+        } else if (criticalRollResult <= 9) {
+            // Damage crit
+            targetUnit.addDamageCrit();
+            reporter.reportDamageCrit(targetUnit);
+        } else if (criticalRollResult <= 11) {
+            // Both targeting and damage crit
+            targetUnit.addTargetingCrit();
+            targetUnit.addDamageCrit();
+            reporter.reportTargetingCrit(targetUnit);
+            reporter.reportDamageCrit(targetUnit);
+        } else {
+            // Destroyed
+            countKill(attackingUnit, targetUnit);
+            targetUnit.setCurrentArmor(0);
+            target.setHighStressEpisode();
+            reporter.reportUnitDestroyed();
+        }
+    }
+
+
+    private void countKill(SBFUnit attackingUnit, SBFUnit targetUnit) {
```

Comment:
CodeQL seems to have a vendetta against this method. I'm leaving it up to you as to whether these are legitimate or false positives.

Reply:
The amount of false positives CodeQL raises makes me wonder how it can be used by anyone.